adam/perception/perception_frame_difference.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC
+from typing import Generic, TypeVar
+
+from attr import attrib, attrs
+from attr.validators import in_, instance_of
+from immutablecollections import ImmutableSet
+from adam.perception.developmental_primitive_perception import DevelopmentalPrimitivePerceptionFrame, ObjectPerception, \
+    PropertyPerception, RelationPerception
+
+
+class PerceptionFrameDiff(ABC):
+    r"""
+    Represents the difference between a `LanguageLearner`\ 's perception
+    of two different moments.
+    """
+
+
+@attrs(slots=True, frozen=True)
+class DevelopmentalPrimitivePerceptionFrameDiff(PerceptionFrameDiff):
+    r"""
+    Represents the difference between a `LanguageLearner`\ 's perception
+    of two different moments as sets of added and removed relations, perceived_objects, and property assertions.
+    """
+
+    added_objects: ImmutableSet["ObjectPerception"] = attrib(validator=instance_of(ImmutableSet))
+    r"""
+    the set of `ObjectPerception`\ s, that were present on the second perception frame, but not the first.
+    """
+    removed_objects: ImmutableSet["ObjectPerception"] = attrib(validator=instance_of(ImmutableSet))
```

Comment:
Consider making all these fields `kw_only=True`, otherwise specifying them as arguments can be error prone